Add result() to ProcessGroupGloo::AsyncWork's

agolynski opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57565 Add result() to ProcessGroupGloo::AsyncWork's**

Differential Revision: [D28255120](https://our.internmc.facebook.com/intern/diff/D28255120)

agolynski committed:
add result() to ProcessGroupGloo::AsyncWork's

[ghstack-poisoned]

agolynski committed:
Update on "WIP. add result() to ProcessGroupGloo::AsyncWork's"




[ghstack-poisoned]

SciPioneer commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -236,8 +236,10 @@ def test_empty_tensors(self):
         xs = [torch.FloatTensor([])]
-        pg.broadcast(xs).wait()
-        self.assertEqual(0, xs[0].numel())
+        w = pg.broadcast(xs)
```
Will the `output` be equal to `xs[0]` after `w.wait()`?

agolynski replied:
added a check

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -418,6 +418,17 @@ const auto kLoopbackAddress = "127.0.0.1";
+  TORCH_CHECK(
+      outputTensors_.size() <= 1,
+      "work result does not support list of list");
+  return outputTensors_.size() == 0 ? std::vector<at::Tensor>() : outputTensors_[0];
```
nit: std::vector<at::Tensor>() can be replaced by "{}"

agolynski replied:
c++ cannot figure out the types in ternary expression, so it complains about {} symbol here

SciPioneer replied:
Interesting to know this, because ideally the compiler should be able to figure it out as you have specified the return type in the function signature.

Out of curiosity, if you do this instead:
`outputTensors_.size() == 1 ? outputTensors_[0] : {};`
Will the compiler be able to figure it out?

I am not saying that you need to change your code though.

SciPioneer commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -190,6 +190,20 @@ std::vector<std::vector<at::Tensor>> copyTensors(
+    } catch (const std::exception& ex) {
+      std::cerr << "Exception received: " << ex.what() << std::endl;
+    }
+    outputTensors.emplace_back(works[i]->result());
```
nit: push_back is better here, because `result()` returns `std::vector<at::Tensor>` instead of `std::vector<at::Tensor>&`.

agolynski replied:
thanks!

yes, moving to outputTensors keeps at least ref to the output tensors. I thought C++ RVO figures this out even for std::vector<at::Tensor>.

efficiency should not be a concern here as it is test?

SciPioneer replied:
Didn't realize it's in the test in my previous quick look. Thx for the udpate!

agolynski replied:
oops, std::move here is not good:

stderr: caffe2/torch/lib/c10d/test/ProcessGroupGlooTest.cpp:202:32: error: moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]
    outputTensors.emplace_back(std::move(works[i]->result()));

https://stackoverflow.com/questions/19267408/why-does-stdmove-prevent-rvo

I think regular RVO should be fine here.

agolynski committed:
Update on "Add result() to ProcessGroupGloo::AsyncWork's"




[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -69,12 +69,16 @@ class ProcessGroupGloo : public ProcessGroup {
   class AsyncWork : public ProcessGroup::Work {
    public:
     AsyncWork(
+        std::vector<std::vector<at::Tensor>> outputTensors,
```
Won't this break backward compatibility?

SciPioneer replied:
Pass by const reference?

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -69,12 +69,16 @@ class ProcessGroupGloo : public ProcessGroup {
   class AsyncWork : public ProcessGroup::Work {
    public:
     AsyncWork(
+        std::vector<std::vector<at::Tensor>> outputTensors,
```
Will this break backward compatibility?

agolynski replied:
no, work constructors aren't part of public facing API.

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -69,12 +69,16 @@ class ProcessGroupGloo : public ProcessGroup {
+        : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
+          outputTensors_(outputTensors) {
     }
 
```
Shouldn't we also override the virtual destructor?

agolynski replied:
if you define virtual destructor in a class it'll be always virtual in subclasses (and will be = default). However, it does not hurt to define one as well.

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -69,12 +69,16 @@ class ProcessGroupGloo : public ProcessGroup {
         const c10::optional<std::vector<at::Tensor>>& inputTensors = c10::nullopt)
-        : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors) {
+        : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
+          outputTensors_(outputTensors) {
```
Shouldn't be `outputTensors_(std::move(outputTensors))`?

Otherwise it will create a separate vector copy. Although the underlying data of each tensor is shared, a copy of this vector can still be non-trivial, unless the length is always short.

agolynski replied:
the intent is to create a separate vector copy, so even if user's output goes out of scope, e.g. user solely relies on result() function as in DDP comm hook code, the underlying tensors won't be going away and this won't result in a SEG.

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -89,8 +93,13 @@ class ProcessGroupGloo : public ProcessGroup {
     friend class ProcessGroupGloo;
+
+   private:
+    std::vector<std::vector<at::Tensor>> outputTensors_;
```
You use `const std::vector<std::vector<at::Tensor>>`, to make sure the length this 2D vector won't change. This will still allow the update of the underlying tensor data.

agolynski replied:
done.

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -69,12 +69,16 @@ class ProcessGroupGloo : public ProcessGroup {
+          outputTensors_(outputTensors) {
     }
 
     static void execute(c10::intrusive_ptr<AsyncWork> work) {
```
nit [optional]: I understand that this part of code is from this PR. Just want to say that since the implementation here is not that trivial, should be moved to cpp file instead.

agolynski replied:
done

agolynski committed:
Update on "Add result() to ProcessGroupGloo::AsyncWork's"




[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28255120).

agolynski committed:
Update on "Add result() to ProcessGroupGloo::AsyncWork's"


Differential Revision: [D28255120](https://our.internmc.facebook.com/intern/diff/D28255120)

[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28255120).

agolynski committed:
Update on "Add result() to ProcessGroupGloo::AsyncWork's"


Differential Revision: [D28255120](https://our.internmc.facebook.com/intern/diff/D28255120)

[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28255120).

agolynski committed:
Update on "Add result() to ProcessGroupGloo::AsyncWork's"


Differential Revision: [D28255120](https://our.internmc.facebook.com/intern/diff/D28255120)

[ghstack-poisoned]

agolynski commented:
Thank you for the quick review @pbelevich @SciPioneer! This is merged now.
